src/box/txn.c

```diff
@@ -988,8 +988,10 @@ txn_prepare(struct txn *txn)
 	 * we have a bunch of IPROTO_NOP statements.
 	 */
 	if (txn->engine != NULL) {
-		if (engine_prepare(txn->engine, txn) != 0)
+		if (engine_prepare(txn->engine, txn) != 0) {
+			txn->psn = 0;
```

Comment:
AFAIU, we use zero psn to distinguish in-progress transactions from prepared ones - wouldn't it be confusing to use zero psn for failed-to-prepare txn?

Reply:
But the point is that a failed-to-prepare transaction's status is not 'prepared' — it's 'in-progress'.

Reply:
Yes, you're right.